pyfifotax/data_structures_event.py

```diff
@@ -0,0 +1,371 @@
+from __future__ import annotations
+
+import decimal
+from datetime import datetime
+from decimal import Decimal
+from typing import Optional
+
+import pandas as pd
+from pandas.core.series import Series
+
+from pyfifotax.data_structures_fifo import FIFOForex, FIFOShare, Forex
+from pyfifotax.data_structures_dataframe import (
+    ESPPRow,
+    RSURow,
+    BuyOrderRow,
+    SellOrderRow,
+    DividendRow,
+    CurrencyConversionRow,
+    StockSplitRow,
+)
+from pyfifotax.utils import to_decimal, round_decimal
+
+
+class ReportEvent:
+    def __init__(self, date: datetime, priority: int):
+        self.date = date
+        # add priority to make sure that buy/deposit transactions
+        # are processed before sell transactions
+        # espp/rsu deposit/dividend: prio 0
+        # buy: prio 1
+        # sell: prio 2
+        # stocksplit: 3 (at the end as assumed after market-close)
+        # currency conversions: 4
+        self.priority = priority
+
+    @staticmethod
+    def from_df_row(row: Series) -> ReportEvent:
+        raise NotImplementedError
+
+    @classmethod
+    def from_report(cls, df: pd.DataFrame) -> list[ReportEvent]:
+        events = []
+        for _, row in df.iterrows():
+            events.append(cls.from_df_row(row))
+        return events
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__} on {self.date.date()}"
+
+
+class RSUEvent(ReportEvent):
+    def __init__(
+        self,
+        date: datetime,
+        symbol: str,
+        received_shares: FIFOShare,
+        withheld_shares: Optional[FIFOShare],
+    ):
+        super().__init__(date, 0)
+        self.symbol = symbol
+        self.received_shares = received_shares
+        self.withheld_shares = withheld_shares
+
+    @staticmethod
+    def from_df_row(row: Series) -> RSUEvent:
+        row = RSURow.from_df_row(row)
+        recv_share = FIFOShare(
+            buy_date=row.date,
+            symbol=row.symbol,
+            buy_price=to_decimal(row.fair_market_value),
```

Comment:
Everywhere in this file with `to_decimal`: "Expected type `Float64Dtype`, got `float` instead"